subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java

```diff
@@ -52,6 +52,21 @@ public interface ProviderFactory {
      *
      * <p>The provider is live and will call the {@link Callable} each time its value is queried. The {@link Callable} may return {@code null}, in which case the provider is considered to have no value.
      *
+     * <h4>Configuration Cache</h4>
+     * <p>This provider is always <a href="provider/Provider.html#configuration-cache">computed and its value is cached</a> by the Configuration Cache.
+     * If this provider is created at configuration time, the {@link Callable} may call configuration-time only APIs and capture objects of arbitrary types.
+     * <p>This can be useful when you need to lazily compute some value to use at execution time based on configuration-time only data. For example, you can compute an archive name based on the name
+     * of the project:
+     * <pre class='autoTested'>
+     *   tasks.register("createArchive") {
+     *       def archiveNameProvider = providers.provider { project.name + ".jar" }
```

Comment:
💭  This is a concise example, but it's not really exemplary, since project name is something that can be operated on at execution time (if extracted first). The `project.name` is artificial in that sense. Maybe there is some other API that is actually configuration-time-only, and extracting which into a variable would not help?